### PR TITLE
[Mono]: Library mode enhancements.

### PR DESCRIPTION
**Pull request by lateralusX**
This PR does a couple of enhancements to library mode builds.

### Add support for direct pinvoke charset on Windows.

NativeAOT direct pinvoke support handles pinvoke charset on Windows:

https://github.com/dotnet/runtime/blob/b69b359b30e0c62099b919024d099a3cb013540d/src/coreclr/tools/aot/ILCompiler/ConfigurablePInvokePolicy.cs#L101

This PR adds the same logic to Mono when targeting Windows, so it will follow the same charset rules as NativeAOT.

### direct pinvoke wrappers on none full AOT builds.

Library mode builds will have option to create a self contained static or shared library, including all needed resources, embedding assemblies and static link all external dependencies (using direct pinvoke). Library mode also exports UnamangedCallersOnly methods that can be called by application loading the library. Currently the wrappers needed for UnmanagedCallersOnly methods will be generated when doing none full AOT builds (https://github.com/dotnet/runtime/commit/741fa51e73e93f3e5752d0d1069a8c871ad302cc), but not pinvoke wrappers. This PR adds support to generate the pinvoke wrappers, but only for the methods described by direct pinvoke AOT compiler arguments, so for none full AOT builds not using direct pinvoke arguments, no pinvoke wrappers will be generated, inline with current none full AOT behavior.

### direct pinvoke wrappers on profiled AOT builds.

Enhanced the profile-only option doing profile AOT to emit managed-to-native and native-to-managed wrappers. Needed in order for library build to work together with profiled AOT builds.

### wrappers-only AOT option.

On platforms like Android, we still run majority of code using JIT. Since library mode should produce a self contained library, including pinvokes (for all direct pinvoked functions) as well as UnmanagedCallersOnly wrappers, Android would need to at least run none full AOT build of all used assemblies. By default, that will AOT a lot more methods that is really needed and there is no way to reduce the AOT:ed code and still run as much as possible using JIT. The profile-only option opens up ability to reduce what managed methods that will be AOT:ed, but still generates a number of inflated methods not needed by UnmanagedCallersOnly and direct pinvoke wrappers and piggy back on an existing option with additional functionality is not optimal. This PR adds an undocumented AOT option intended to be used by library builder, `wrappers-only`, and that will only include wrappers in AOT object giving library builder the ability to build a self contained library on Android that still runs majority of managed code using JIT.

### CODE REVIEW

File: src/mono/mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -13886,12 +14022,12 @@ add_mibc_profile_methods (MonoAotCompile *acfg, char *filename)
...
 free_method_pinvoke_import_value (gpointer data)
 {
-	gchar **value = (gchar **)data;
+	PInvokeImportData *value = (PInvokeImportData *)data;
 	if (!value)
 		return;
-	g_free (value[0]);
-	g_free (value[1]);
-	g_free (value);
+	g_free ((char *)(value->module));
```
Comment (vargaz) [inline src/mono/mono/mini/aot-compiler.c:14028]: Are these casts needed ?

File: src/mono/mono/mini/aot-compiler.c [hunk 2]
```diff
@@ -256,6 +256,7 @@ typedef struct MonoAotOptions {
 	char *clangxx;
 	char *depfile;
 	char *runtime_init_callback;
+	gboolean wrappers_only;
```
Comment (vargaz) [inline src/mono/mono/mini/aot-compiler.c:259]: Move this gboolean to be with the others ?

File: src/mono/mono/mini/aot-compiler.c [hunk 3]
```diff
@@ -5154,6 +5176,37 @@ add_wrappers (MonoAotCompile *acfg)
 	}
 }
 
+static void
+add_managed_to_native_wrappers (MonoAotCompile *acfg)
+{
```
Comment (vargaz) [inline src/mono/mono/mini/aot-compiler.c:5181]: Shouldn't add_wrappers () call this ?

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/mini/aot-compiler.c (hunk 1 above)
Comment (lateralusX) [reply src/mono/mono/mini/aot-compiler.c:14028]: Not really needed, could be removed, or stay.

File: src/mono/mono/mini/aot-compiler.c (hunk 3 above)
Comment (lateralusX) [reply src/mono/mono/mini/aot-compiler.c:5181]: add_wrappers is only called from one location, and that is where we now also calls add_managed_to_native_wrappers as well as add_native_to_managed_wrappers. I initially moved all into add_wrappers and added condition in add wrappers to only call the wrappers under full aot, but that diff caused a lot of noise to the overall PR, but let me try that again and see if I can get the diff a little cleaner.

**Comment by lateralusX** [comment]
Failures are known issues.

**Pull request merged**